src/GitHub.Api/Installer/GitInstaller.cs

```diff
@@ -116,45 +117,84 @@ public void SetupGitIfNeeded(ActionTask<NPath> onSuccess, ITask onFailure)
                 return;
             }
 
-            new ActionTask(cancellationToken, () => {
-                if (IsGitExtracted())
+            var isGitExtractedTask = new FuncTask<NPath>(cancellationToken, () =>
+            {
+                if (!IsGitExtracted())
                 {
-                    Logger.Trace("SetupGitIfNeeded: Skipped");
-                    onSuccess.PreviousResult = installDetails.GitExecutablePath;
-                    onSuccess.Start();
+                    GrabZipFromResources();
+                    return null;
                 }
-                else
+                Logger.Trace("SetupGitIfNeeded: Skipped");
+                return installDetails.GitExecutablePath;
+            });
+            isGitExtractedTask.OnEnd += (t, res, _, __) =>
+            {
+                if (res == null)
                 {
-                    ExtractPortableGit(onSuccess, onFailure);
+                    var extractTask = ExtractPortableGit();
+                    extractTask.Then(onSuccess, TaskRunOptions.OnSuccess, taskIsTopOfChain: true);
+                    extractTask.Then(onFailure, TaskRunOptions.OnFailure, taskIsTopOfChain: true);
+                    t.Then(extractTask);
```

Comment:
If this `isGitExtractedTask` has already ended. How is this different from just starting extractTask?